ml4gw/waveforms/phenom_p.py

```diff
@@ -0,0 +1,731 @@
+        tmp_x, tmp_y, tmp_z = 0, 0, 1
+        tmp_x, tmp_y, tmp_z = self.ROTATEZ(-phiJ_sf, tmp_x, tmp_y, tmp_z)
+        tmp_x, tmp_y, tmp_z = self.ROTATEY(-thetaJ_sf, tmp_x, tmp_y, tmp_z)
+        tmp_x, tmp_y, tmp_z = self.ROTATEZ(kappa, tmp_x, tmp_y, tmp_z)
+
+        alpha0 = torch.arctan2(tmp_y, tmp_x)
+
+        # Finally we determine thetaJ, by rotating N
+        tmp_x, tmp_y, tmp_z = Nx_sf, Ny_sf, Nz_sf
+        tmp_x, tmp_y, tmp_z = self.ROTATEZ(-phiJ_sf, tmp_x, tmp_y, tmp_z)
+        tmp_x, tmp_y, tmp_z = self.ROTATEY(-thetaJ_sf, tmp_x, tmp_y, tmp_z)
+        tmp_x, tmp_y, tmp_z = self.ROTATEZ(kappa, tmp_x, tmp_y, tmp_z)
+        Nx_Jf, Nz_Jf = tmp_x, tmp_z
+        thetaJN = torch.arccos(Nz_Jf)
+
+        # Finally, we need to redefine the polarizations:
+        # PhenomP's polarizations are defined following Arun et al
+        # (arXiv:0810.5336)
+        # i.e. projecting the metric onto the P,Q,N triad defined with
+        # P=NxJ/|NxJ| (see (2.6) in there).
+        # By contrast, the triad X,Y,N used in LAL
+        # ("waveframe" in the nomenclature of T1500606-v6)
+        # is defined in e.g. eq (35) of this document
+        # (via its components in the source frame;
+        # note we use the default Omega=Pi/2).
+        # Both triads differ from each other by a rotation around N by an angle
+        # \zeta and we need to rotate the polarizations accordingly by 2\zeta
+
+        Xx_sf = -torch.cos(incl) * torch.sin(phiRef)
+        Xy_sf = -torch.cos(incl) * torch.cos(phiRef)
+        Xz_sf = torch.sin(incl)
+        tmp_x, tmp_y, tmp_z = Xx_sf, Xy_sf, Xz_sf
+        tmp_x, tmp_y, tmp_z = self.ROTATEZ(-phiJ_sf, tmp_x, tmp_y, tmp_z)
+        tmp_x, tmp_y, tmp_z = self.ROTATEY(-thetaJ_sf, tmp_x, tmp_y, tmp_z)
+        tmp_x, tmp_y, tmp_z = self.ROTATEZ(kappa, tmp_x, tmp_y, tmp_z)
+
+        # Now the tmp_a are the components of X in the J frame
+        # We need the polar angle of that vector in the P,Q basis of Arun et al
+        # P = NxJ/|NxJ| and since we put N in the (pos x)z half plane of the J
+        # frame
+        PArunx_Jf = 0.0
+        PAruny_Jf = -1.0
+        PArunz_Jf = 0.0
+
+        # Q = NxP
+        QArunx_Jf = Nz_Jf
+        QAruny_Jf = 0.0
+        QArunz_Jf = -Nx_Jf
+
+        # Calculate the dot products XdotPArun and XdotQArun
+        XdotPArun = tmp_x * PArunx_Jf + tmp_y * PAruny_Jf + tmp_z * PArunz_Jf
+        XdotQArun = tmp_x * QArunx_Jf + tmp_y * QAruny_Jf + tmp_z * QArunz_Jf
+
+        zeta_polariz = torch.arctan2(XdotQArun, XdotPArun)
+        return chi1_l, chi2_l, chip, thetaJN, alpha0, phi_aligned, zeta_polariz
+
+    # TODO: add input and output types
+    def SpinWeightedY(self, theta, phi, s, l, m):  # noqa: E741
+        "copied from SphericalHarmonics.c in LAL"
+        if s == -2:
+            if l == 2:  # noqa: E741
+                if m == -2:
+                    fac = (
+                        torch.sqrt(torch.tensor(5.0 / (64.0 * PI)))
```

Comment:
Have you checked if this works when the IMRPhenomPv2 instance is sent to the GPU? The `torch.tensor(5.0 / (64.0 * PI)` here and other similar instances will define the quantity on the CPU, hence asking.

Reply:
Also generally if we can avoid the `if...else` conditionals that would be even better. Can we register the Ylm factors as buffers and evaluate the Ylms as needed?

Reply:
I dont think theres any way around it, since the factor `fac` depends on thetaJ, which depends on the input params.